File: LeetCode-go/cn/wsharkcoder/leetcode/P1.TwoSum.go
package leetcode

// 解法一、双循环暴力解 O(n^2)
//func twoSum(nums []int, target int) []int {
//	var result []int = make([]int, 2)
//	for i := 0; i < len(nums)-1; i++ {
//		for j := i + 1; j < len(nums); j++ {
//			if nums[i]+nums[j] == target {
//				result[0] = i
//				result[1] = j
//				break
//			}
//		}
//	}
//	return result
//}

// 解法二、一遍遍历数组，一遍往map中放，然后判断另外一个目标值在不在map中 O(n)
func twoSum(nums []int, target int) []int {

	m := make(map[int]int, len(nums))

	for j := 0; j < len(nums); j++ {
		another := target - nums[j]
		// Go语言map特性,同时获取值和是否有值
		i, ok := m[another]
		if ok {
			return []int{i, j}

		}
		m[nums[j]] = j
	}
	return nil
}

File: LeetCode-Java/src/main/java/P1TwoSum.java
import java.util.HashMap;

/**
 * 1.两数之和
 *
 * @author WSharkCoder
 * @date 2022/03/07
 */
public class P1TwoSum {

    static class Solution {
        /**
         * 解题思路:
         * 倒序索引 map<value,index>，遍历每个数据，判断另外一个（相加=target）值是否在倒序索引中
         * 时间复杂度: O(n)
         *
         * @param nums   数据数组
         * @param target 目标值
         * @return 下标
         */
        public int[] twoSum(int[] nums, int target) {
            // val->index映射
            HashMap<Integer, Integer> map = new HashMap<>();
            for (int i = 0; i < nums.length; i++) {
                if (map.containsKey(target - nums[i])) {
                    return new int[]{map.get(target - nums[i]), i};
                }
                map.put(nums[i], i);
            }
            return null;
        }
    }
}
